File: examples/kotlin/examples/ClientKvPrecision.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.*

private fun unwrapField(field: JsonElement?): JsonElement? =
    if (field is JsonObject && "type" in field && "value" in field) field["value"] else field

private fun fieldDouble(field: JsonElement?): Double? =
    (unwrapField(field) as? JsonPrimitive)?.doubleOrNull

private fun fieldString(field: JsonElement?): String? =
    (unwrapField(field) as? JsonPrimitive)?.contentOrNull

private fun kvObject(result: Any?): JsonObject {
    var value = result as? JsonElement ?: error("KV response was not JSON: $result")
    repeat(3) {
        value = when {
            value is JsonPrimitive && value.isString ->
                Json.parseToJsonElement(value.content)
            value is JsonObject && "type" in value && "value" in value ->
                value["value"] ?: error("KV response value was null")
            value is JsonObject && value.size == 1 && "value" in value ->
                value["value"] ?: error("KV response value was null")
            else -> return@repeat
        }
    }
    return value as? JsonObject ?: error("KV response did not contain an object: $result")
}

/**
 * KV Precision example - Float vs Decimal Comparison
 *
 * Demonstrates the difference between using Kotlin Doubles (which lose precision)
 * vs ekoDB's Decimal type (which preserves precision).
 *
 * Key takeaway: Use fieldDecimal() for monetary values and other cases
 * where precision matters.
 */
fun main() = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"

    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()

    println("=== ekoDB Kotlin Client - KV Precision: Float vs Decimal ===\n")

    // Track all keys for cleanup
    val allKeys = mutableListOf<String>()
    var primaryError: Throwable? = null

    try {
        // =====================================================================
        // Test 1: Float Precision Loss
        // =====================================================================
        println("=== Test 1: Using Kotlin Doubles (LOSES PRECISION) ===")

        val floatProducts = listOf(
            Triple("product:float:1", buildJsonObject { put("name", "Widget A"); put("price", 29.99) }, null),
            Triple("product:float:2", buildJsonObject { put("name", "Widget B"); put("price", 39.99) }, null),
            Triple("product:float:3", buildJsonObject { put("name", "Widget C"); put("price", 49.99) }, null)
        )
        val floatKeys = floatProducts.map { it.first }
        allKeys.addAll(floatKeys)

        // Store with float values
        client.kvBatchSet(floatProducts)
        println("Stored products with float prices")

        // Retrieve and show the precision loss
        val floatResults = client.kvBatchGet(floatKeys)
        check(floatResults.size == floatProducts.size) {
            "Expected ${floatProducts.size} float products, got ${floatResults.size}"
        }
        println("\nRetrieved float prices:")
        val expectedFloatPrices = listOf(29.99, 39.99, 49.99)
        val productNames = listOf("Widget A", "Widget B", "Widget C")
        for (i in floatResults.indices) {
            val obj = kvObject(floatResults[i])
            val actualPrice = fieldDouble(obj["price"])
                ?: error("Missing float price for ${productNames[i]}")
            val expectedPrice = expectedFloatPrices[i]
            val match = if (expectedPrice == actualPrice) "MATCH" else "PRECISION LOST"
            println("  ${productNames[i]}: \$$actualPrice (expected \$$expectedPrice) $match")
            check(actualPrice == expectedPrice) {
                "Float price for ${productNames[i]} was $actualPrice, expected $expectedPrice"
            }
        }

        // =====================================================================
        // Test 2: Decimal Precision Preserved
        // =====================================================================
        println("\n=== Test 2: Using fieldDecimal() (PRESERVES PRECISION) ===")

        // fieldDecimal() returns Map<String, Any>, but kvBatchSet needs JsonElement values.
        // Build the JSON objects with decimal type wrappers manually.
        val decimalProducts = listOf(
            Triple("product:decimal:1", buildJsonObject {
                put("name", "Widget A")
                putJsonObject("price") { put("type", "Decimal"); put("value", "29.99") }
            }, null),
            Triple("product:decimal:2", buildJsonObject {
                put("name", "Widget B")
                putJsonObject("price") { put("type", "Decimal"); put("value", "39.99") }
            }, null),
            Triple("product:decimal:3", buildJsonObject {
                put("name", "Widget C")
                putJsonObject("price") { put("type", "Decimal"); put("value", "49.99") }
            }, null)
        )
        val decimalKeys = decimalProducts.map { it.first }
        allKeys.addAll(decimalKeys)

        // Store with decimal values
        client.kvBatchSet(decimalProducts)
        println("Stored products with decimal prices")

        // Retrieve and show precision is preserved
        val decimalResults = client.kvBatchGet(decimalKeys)
        check(decimalResults.size == decimalProducts.size) {
            "Expected ${decimalProducts.size} decimal products, got ${decimalResults.size}"
        }
        val expectedDecimalPrices = listOf("29.99", "39.99", "49.99")
        println("\nRetrieved decimal prices:")
        for (i in decimalResults.indices) {
            val obj = kvObject(decimalResults[i])
            val actualPrice = fieldString(obj["price"])
                ?: error("Missing decimal price for ${productNames[i]}")
            val expectedPrice = expectedDecimalPrices[i]
            check(actualPrice == expectedPrice) {
                "Decimal price for ${productNames[i]} was $actualPrice, expected $expectedPrice"
            }
            println("  ${productNames[i]}: \$$actualPrice (expected \$$expectedPrice)")
        }

        // =====================================================================
        // Test 3: Real-world calculation comparison
        // =====================================================================
        println("\n=== Test 3: Sum Calculation Comparison ===")

        // Sum up float prices
        var floatSum = 0.0
        for ((index, result) in floatResults.withIndex()) {
            val price = fieldDouble(kvObject(result)["price"])
                ?: error("Missing float price for ${productNames[index]}")
            floatSum += price
        }
        println("  Float sum: \$$floatSum (expected \$119.97)")
        check(kotlin.math.abs(floatSum - 119.97) < 1e-9) {
            "Float sum was $floatSum, expected 119.97"
        }

        // Sum up decimal prices (come back as strings, convert for display)
        var decimalSum = java.math.BigDecimal.ZERO
        for ((index, result) in decimalResults.withIndex()) {
            val price = fieldString(kvObject(result)["price"])
                ?: error("Missing decimal price for ${productNames[index]}")
            decimalSum = decimalSum.add(price.toBigDecimal())
        }
        check(decimalSum == "119.97".toBigDecimal()) {
            "Decimal sum was $decimalSum, expected 119.97"
        }
        println("  Decimal sum: \$${decimalSum.toPlainString()} (expected \$119.97)")

        // =====================================================================
        // Test 4: More extreme precision example
        // =====================================================================
        println("\n=== Test 4: Extreme Precision Example ===")

        // Store a value that floats can't represent exactly: 0.1 + 0.2
        val floatPrecisionKey = "precision:float"
        val decimalPrecisionKey = "precision:decimal"
        allKeys.add(floatPrecisionKey)
        allKeys.add(decimalPrecisionKey)

        client.kvSet(floatPrecisionKey, buildJsonObject {
            put("amount", 0.1 + 0.2) // Classic floating-point imprecision
        })
        client.kvSet(decimalPrecisionKey, buildJsonObject {
            putJsonObject("amount") { put("type", "Decimal"); put("value", "0.30") }
        })

        // Retrieve and compare
        val floatPrecision = client.kvGet(floatPrecisionKey)
        val decimalPrecision = client.kvGet(decimalPrecisionKey)

        val floatParsed = kvObject(floatPrecision)
        val decimalParsed = kvObject(decimalPrecision)

        val floatAmount = fieldDouble(floatParsed["amount"])
            ?: error("Missing float precision amount in $floatPrecision")
        val decimalAmount = fieldString(decimalParsed["amount"])
            ?: error("Missing decimal precision amount in $decimalPrecision")
        println("  Float 0.1 + 0.2 = $floatAmount (should be 0.3)")
        println("  Decimal \"0.30\" = $decimalAmount (exact!)")
        check(floatAmount == 0.1 + 0.2) {
            "Float precision amount was $floatAmount, expected ${0.1 + 0.2}"
        }
        check(decimalAmount == "0.30") {
            "Decimal precision amount was $decimalAmount, expected 0.30"
        }

    } catch (error: Throwable) {
        primaryError = error
        throw error
    } finally {
        // =====================================================================
        // Cleanup
        // =====================================================================
        println("\n=== Cleanup ===")
        var cleanupError: Throwable? = null
        try {
            client.kvBatchDelete(allKeys)
            println("Cleaned up ${allKeys.size} test keys")
        } catch (error: Throwable) {
            println("Cleanup error: ${error.message}")
            cleanupError = error
        }

        try {
            client.close()
        } catch (error: Throwable) {
            if (cleanupError == null) {
                cleanupError = error
            } else {
                cleanupError.addSuppressed(error)
            }
        }

        cleanupError?.let { error ->
            if (primaryError == null) throw error
            primaryError.addSuppressed(error)
        }
    }

    println("\n=== Summary ===")
    println("Use fieldDecimal() for monetary values, percentages, and")
    println("any case where floating-point errors are unacceptable.")
    println("fieldDecimal() stores values as strings internally,")
    println("preserving exact precision across all operations.")
    println("\n=== Example Complete ===")
}

File: examples/rust/examples/client_kv_precision.rs
//! KV Precision Example - Float vs Decimal Comparison
//!
//! This example demonstrates the difference between using f64 floats
//! (which can lose precision) vs ekoDB's Decimal type (which preserves precision).
//!
//! Key takeaway: Use FieldType::decimal() for monetary values and other cases
//! where precision matters.

use ekodb_client::{get_string_value, get_value, Client, FieldType, Record};
use rust_decimal::Decimal;
use serde_json::{json, Value};
use std::env;
use std::str::FromStr;

const FLOAT_KEYS: [&str; 3] = [
    "kv_precision:rs:product:float:1",
    "kv_precision:rs:product:float:2",
    "kv_precision:rs:product:float:3",
];
const DECIMAL_KEYS: [&str; 3] = [
    "kv_precision:rs:product:decimal:1",
    "kv_precision:rs:product:decimal:2",
    "kv_precision:rs:product:decimal:3",
];
const FLOAT_PRECISION_KEY: &str = "kv_precision:rs:precision:float";
const DECIMAL_PRECISION_KEY: &str = "kv_precision:rs:precision:decimal";

fn all_keys() -> Vec<String> {
    FLOAT_KEYS
        .into_iter()
        .chain(DECIMAL_KEYS)
        .chain([FLOAT_PRECISION_KEY, DECIMAL_PRECISION_KEY])
        .map(str::to_string)
        .collect()
}

fn kv_object(value: Value, label: &str) -> Result<Value, Box<dyn std::error::Error>> {
    let decoded = match value {
        Value::String(raw) => serde_json::from_str(&raw)?,
        value => value,
    };
    let unwrapped = get_value(&decoded);
    if !unwrapped.is_object() {
        return Err(format!("{label} KV value was not an object: {unwrapped}").into());
    }
    Ok(unwrapped)
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Create ekoDB client
    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    client.kv_batch_delete(all_keys()).await?;

    let run_result: Result<(), Box<dyn std::error::Error>> = async {
        println!("=== KV Precision: Float vs Decimal ===\n");

        // =========================================================================
        // Test 1: Float Precision Loss
        // =========================================================================
        println!("=== Test 1: Using f64 Floats (MAY LOSE PRECISION) ===");

        let float_keys = FLOAT_KEYS.map(str::to_string).to_vec();
        let float_prices = [29.99_f64, 39.99, 49.99];
        let float_names = ["Widget A", "Widget B", "Widget C"];

        // Store with float values
        let mut float_values = Vec::new();
        for i in 0..3 {
            let mut record = Record::new();
            record.insert("name", float_names[i]);
            record.insert("price", float_prices[i]);
            float_values.push(record);
        }
        client
            .kv_batch_set(float_keys.clone(), float_values, None)
            .await?;
        println!("Stored products with float prices");

        // Retrieve and show the precision
        let float_results = client.kv_batch_get(float_keys.clone()).await?;
        if float_results.len() != float_prices.len() {
            return Err(format!(
                "expected {} float products, got {}",
                float_prices.len(),
                float_results.len()
            )
            .into());
        }
        println!("\nRetrieved float prices:");
        for (i, record) in float_results.iter().enumerate() {
            let price_field = record
                .get("price")
                .ok_or_else(|| format!("missing float price for {}", float_names[i]))?;
            let price_val = get_value(&json!(price_field));
            let actual_price = price_val
                .as_f64()
                .ok_or_else(|| format!("float price for {} was not numeric", float_names[i]))?;
            let expected = float_prices[i];
            let matches = (expected - actual_price).abs() < f64::EPSILON;
            let status = if matches { "OK" } else { "PRECISION LOST" };
            println!(
                "  {}: ${} (expected ${}) {}",
                float_names[i], actual_price, expected, status
            );
            if !matches {
                return Err(format!(
                    "unexpected float price for {}: {}",
                    float_names[i], actual_price
                )
                .into());
            }
        }

        // =========================================================================
        // Test 2: Decimal Precision Preserved
        // =========================================================================
        println!("\n=== Test 2: Using FieldType::decimal() (PRESERVES PRECISION) ===");

        let decimal_keys = DECIMAL_KEYS.map(str::to_string).to_vec();
        let decimal_prices = ["29.99", "39.99", "49.99"];

        // Store with decimal values
        let mut decimal_values = Vec::new();
        for i in 0..3 {
            let mut record = Record::new();
            record.insert("name", float_names[i]);
            record.insert(
                "price",
                FieldType::decimal(Decimal::from_str(decimal_prices[i])?),
            );
            decimal_values.push(record);
        }
        client
            .kv_batch_set(decimal_keys.clone(), decimal_values, None)
            .await?;
        println!("Stored products with decimal prices");

        // Retrieve and show precision is preserved
        // Decimal values come back as {"type": "Decimal", "value": "29.99"} - get_string_value extracts the string
        let decimal_results = client.kv_batch_get(decimal_keys.clone()).await?;
        if decimal_results.len() != decimal_prices.len() {
            return Err(format!(
                "expected {} decimal products, got {}",
                decimal_prices.len(),
                decimal_results.len()
            )
            .into());
        }
        println!("\nRetrieved decimal prices:");
        for (i, record) in decimal_results.iter().enumerate() {
            let price_field = record
                .get("price")
                .ok_or_else(|| format!("missing decimal price for {}", float_names[i]))?;
            let actual_str = get_string_value(&json!(price_field))
                .ok_or_else(|| format!("decimal price for {} was not a string", float_names[i]))?;
            println!(
                "  {}: ${} (expected ${}) OK",
                float_names[i], actual_str, decimal_prices[i]
            );
            if actual_str != decimal_prices[i] {
                return Err(format!(
                    "unexpected decimal price for {}: {}",
                    float_names[i], actual_str
                )
                .into());
            }
        }

        // =========================================================================
        // Test 3: Real-world calculation comparison
        // =========================================================================
        println!("\n=== Test 3: Sum Calculation Comparison ===");

        // Sum float prices
        let mut float_sum = 0.0_f64;
        for (i, record) in float_results.iter().enumerate() {
            let price_field = record
                .get("price")
                .ok_or_else(|| format!("missing float price for {}", float_names[i]))?;
            let price_val = get_value(&json!(price_field));
            float_sum += price_val
                .as_f64()
                .ok_or_else(|| format!("float price for {} was not numeric", float_names[i]))?;
        }
        println!("  Float sum: ${} (expected $119.97)", float_sum);
        if (float_sum - 119.97).abs() > 1e-9 {
            return Err(format!("unexpected float sum: {float_sum}").into());
        }

        // Sum decimal prices (values come back as strings, parse for arithmetic)
        let mut decimal_sum = Decimal::ZERO;
        for (i, record) in decimal_results.iter().enumerate() {
            let price_field = record
                .get("price")
                .ok_or_else(|| format!("missing decimal price for {}", float_names[i]))?;
            let price_str = get_string_value(&json!(price_field))
                .ok_or_else(|| format!("decimal price for {} was not a string", float_names[i]))?;
            decimal_sum += Decimal::from_str(&price_str)?;
        }
        println!("  Decimal sum: ${} (expected $119.97)", decimal_sum);
        if decimal_sum != Decimal::from_str("119.97")? {
            return Err(format!("unexpected decimal sum: {decimal_sum}").into());
        }

        // =========================================================================
        // Test 4: Extreme precision example
        // =========================================================================
        println!("\n=== Test 4: Extreme Precision Example ===");

        // Store 0.1 + 0.2 as a float (classic floating-point issue)
        client
            .kv_set(
                FLOAT_PRECISION_KEY,
                json!({"amount": 0.1_f64 + 0.2_f64}),
                None,
            )
            .await?;

        // Store "0.30" as a decimal (exact representation)
        client
            .kv_set(
                DECIMAL_PRECISION_KEY,
                json!({"amount": FieldType::decimal(Decimal::from_str("0.30")?)}),
                None,
            )
            .await?;

        // Retrieve and compare
        let float_val = client
            .kv_get(FLOAT_PRECISION_KEY)
            .await?
            .ok_or("precision:float was not returned")?;
        let float_object = kv_object(float_val, "precision:float")?;
        let float_amount = float_object
            .get("amount")
            .map(get_value)
            .and_then(|value| value.as_f64())
            .ok_or("precision:float amount was not numeric")?;
        println!("  Float 0.1 + 0.2 = {:.17} (should be 0.3)", float_amount);
        if float_amount != 0.1_f64 + 0.2_f64 {
            return Err(format!("unexpected float precision value: {float_amount}").into());
        }

        let decimal_val = client
            .kv_get(DECIMAL_PRECISION_KEY)
            .await?
            .ok_or("precision:decimal was not returned")?;
        let decimal_object = kv_object(decimal_val, "precision:decimal")?;
        let decimal_amount = decimal_object
            .get("amount")
            .and_then(get_string_value)
            .ok_or("precision:decimal amount was not a string")?;
        println!("  Decimal \"0.30\" = {} (exact!)", decimal_amount);
        if decimal_amount != "0.30" {
            return Err(format!("unexpected decimal precision value: {decimal_amount}").into());
        }

        Ok(())
    }
    .await;

    // =========================================================================
    // Cleanup
    // =========================================================================
    println!("\n=== Cleanup ===");
    let cleanup_result = client.kv_batch_delete(all_keys()).await;
    if cleanup_result.is_ok() {
        println!("Cleaned up test keys");
    }

    match (run_result, cleanup_result) {
        (Err(primary), Err(cleanup)) => {
            return Err(format!("{primary}; cleanup also failed: {cleanup}").into());
        }
        (Err(primary), Ok(_)) => return Err(primary),
        (Ok(_), Err(cleanup)) => return Err(cleanup.into()),
        (Ok(_), Ok(_)) => {}
    }

    println!("\n=== Summary ===");
    println!("Use FieldType::decimal() for monetary values, percentages, and");
    println!("any case where floating-point errors are unacceptable.");
    println!("Decimal values are stored as exact strings internally,");
    println!("preserving precision across all operations.");

    Ok(())
}
